<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/images/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>

<h1 align=center><font size = 5>Assignment: Notebook for Graded Assessment</font></h1>


# Introduction

Using this Python notebook you will:

1.  Understand three Chicago datasets
2.  Load the three datasets into three tables in a SQLIte database
3.  Execute SQL queries to answer assignment questions


## Understand the datasets

To complete the assignment problems in this notebook you will be using three datasets that are available on the city of Chicago's Data Portal:

1.  <a href="https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Socioeconomic Indicators in Chicago</a>
2.  <a href="https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Public Schools</a>
3.  <a href="https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Crime Data</a>

### 1. Socioeconomic Indicators in Chicago

This dataset contains a selection of six socioeconomic indicators of public health significance and a “hardship index,” for each Chicago community area, for the years 2008 – 2012.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2](https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 2. Chicago Public Schools

This dataset shows all school level performance data used to create CPS School Report Cards for the 2011-2012 school year. This dataset is provided by the city of Chicago's Data Portal.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 3. Chicago Crime Data

This dataset reflects reported incidents of crime (with the exception of murders where data exists for each victim) that occurred in the City of Chicago from 2001 to present, minus the most recent seven days.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)


### Download the datasets

This assignment requires you to have these three tables populated with a subset of the whole datasets.

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. 

Use the links below to read the data files using the Pandas library. 

* Chicago Census Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Public Schools

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Crime Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

**NOTE:** Ensure you use the datasets available on the links above instead of directly from the Chicago Data Portal. The versions linked here are subsets of the original datasets and have some of the column names modified to be more database friendly which will make it easier to complete this assignment.


Execute the below code cell to install the required libraries


In [1]:
!pip install pandas
!pip install ipython-sql prettytable 

import prettytable

prettytable.DEFAULT = 'DEFAULT'

### Store the datasets in database tables

To analyze the data using SQL, it first needs to be loaded into SQLite DB.
We will create three tables in as under:

1.  **CENSUS_DATA**
2.  **CHICAGO_PUBLIC_SCHOOLS**
3.  **CHICAGO_CRIME_DATA**


Load the `pandas` and `sqlite3` libraries and establish a connection to `FinalDB.db`


In [3]:
# Install required packages
!pip install sqlalchemy==1.4.53
!pip install -q pandas==2.2.3
!pip install ipython-sql==0.5.0
!pip install requests

# Load or reload the SQL extension
%reload_ext sql

# Connect to SQLite database
%sql sqlite:///FinalDB.db

import pandas as pd
import sqlite3
import csv
import requests
import io

# Function to fetch CSV with SSL verification disabled
def fetch_csv(url):
    # Disable SSL verification (use with caution)
    response = requests.get(url, verify=False)
    response.raise_for_status()  # Raise an error for bad responses
    return pd.read_csv(io.StringIO(response.text))

# Create SQLite connection
con = sqlite3.connect("FinalDB.db")

# Load and write Census Data
df = fetch_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv")
df.to_sql("CENSUS_DATA", con, if_exists='replace', index=False, chunksize=100)

# Load and write Chicago Public Schools Data
df = fetch_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv")
df.to_sql("CHICAGO_PUBLIC_SCHOOLS", con, if_exists='replace', index=False, chunksize=100)

# Load and write Chicago Crime Data
df = fetch_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv")
df.to_sql("CHICAGO_CRIME_DATA", con, if_exists='replace', index=False, chunksize=100)

# Close connection
con.close()

# Run your query
%sql SELECT COUNT(*) FROM CHICAGO_CRIME_DATA

  Using cached SQLAlchemy-1.4.53-cp311-cp311-macosx_10_9_universal2.whl.metadata (10 kB)
Using cached SQLAlchemy-1.4.53-cp311-cp311-macosx_10_9_universal2.whl (1.6 MB)
  Attempting uninstall: sqlalchemy
    Found existing installation: SQLAlchemy 2.0.41
    Uninstalling SQLAlchemy-2.0.41:
      Successfully uninstalled SQLAlchemy-2.0.41
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
ipython-sql 0.5.0 requires sqlalchemy>=2.0, but you have sqlalchemy 1.4.53 which is incompatible.
  Using cached sqlalchemy-2.0.41-cp311-cp311-macosx_10_9_x86_64.whl.metadata (9.6 kB)
Using cached sqlalchemy-2.0.41-cp311-cp311-macosx_10_9_x86_64.whl (2.1 MB)
  Attempting uninstall: sqlalchemy
    Found existing installation: SQLAlchemy 1.4.53
    Uninstalling SQLAlchemy-1.4.53:
      Successfully uninstalled SQLAlchemy-1.4.53


/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/urllib3/connectionpool.py:1097: InsecureRequestWarning: Unverified HTTPS request is being made to host 'cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/urllib3/connectionpool.py:1097: InsecureRequestWarning: Unverified HTTPS request is being made to host 'cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/urllib3/connectionpool.py:1097: InsecureRequestWarning: Unverified HTTPS request is being made to host 'cf-courses-data.s3.us.cloud-

 * sqlite:///FinalDB.db
Done.


[(533,)]

Load the SQL magic module


In [4]:
con = sqlite3.connect("FinalDB.db")
cur = con.cursor()

Use `Pandas` to load the data available in the links above to dataframes. Use these dataframes to load data on to the database `FinalDB.db` as required tables.


Establish a connection between SQL magic module and the database `FinalDB.db`


You can now proceed to the the following questions. Please note that a graded assignment will follow this lab and there will be a question on each of the problems stated below. It can be from the answer you received or the code you write for this problem. Therefore, please keep a note of both your codes as well as the response you generate.


## Problems

Now write and execute SQL queries to solve assignment problems

### Problem 1

##### Find the total number of crimes recorded in the CRIME table.


In [5]:
%sql SELECT COUNT(*) FROM CHICAGO_CRIME_DATA

 * sqlite:///FinalDB.db
Done.


[(533,)]

### Problem 2

##### List community area names and numbers with per capita income less than 11000.


In [6]:
%sql SELECT COMMUNITY_AREA_NAME, COMMUNITY_AREA_NUMBER FROM CENSUS_DATA WHERE PER_CAPITA_INCOME	< 110000

 * sqlite:///FinalDB.db
Done.


[('Rogers Park', 1.0),
 ('West Ridge', 2.0),
 ('Uptown', 3.0),
 ('Lincoln Square', 4.0),
 ('North Center', 5.0),
 ('Lake View', 6.0),
 ('Lincoln Park', 7.0),
 ('Near North Side', 8.0),
 ('Edison Park', 9.0),
 ('Norwood Park', 10.0),
 ('Jefferson Park', 11.0),
 ('Forest Glen', 12.0),
 ('North Park', 13.0),
 ('Albany Park', 14.0),
 ('Portage Park', 15.0),
 ('Irving Park', 16.0),
 ('Dunning', 17.0),
 ('Montclaire', 18.0),
 ('Belmont Cragin', 19.0),
 ('Hermosa', 20.0),
 ('Avondale', 21.0),
 ('Logan Square', 22.0),
 ('Humboldt park', 23.0),
 ('West Town', 24.0),
 ('Austin', 25.0),
 ('West Garfield Park', 26.0),
 ('East Garfield Park', 27.0),
 ('Near West Side', 28.0),
 ('North Lawndale', 29.0),
 ('South Lawndale', 30.0),
 ('Lower West Side', 31.0),
 ('Loop', 32.0),
 ('Near South Side', 33.0),
 ('Armour Square', 34.0),
 ('Douglas', 35.0),
 ('Oakland', 36.0),
 ('Fuller Park', 37.0),
 ('Grand Boulevard', 38.0),
 ('Kenwood', 39.0),
 ('Washington Park', 40.0),
 ('Hyde Park', 41.0),
 ('Woodlawn', 42.0),
 ('South Shore', 43.0),
 ('Chatham', 44.0),
 ('Avalon Park', 45.0),
 ('South Chicago', 46.0),
 ('Burnside', 47.0),
 ('Calumet Heights', 48.0),
 ('Roseland', 49.0),
 ('Pullman', 50.0),
 ('South Deering', 51.0),
 ('East Side', 52.0),
 ('West Pullman', 53.0),
 ('Riverdale', 54.0),
 ('Hegewisch', 55.0),
 ('Garfield Ridge', 56.0),
 ('Archer Heights', 57.0),
 ('Brighton Park', 58.0),
 ('McKinley Park', 59.0),
 ('Bridgeport', 60.0),
 ('New City', 61.0),
 ('West Elsdon', 62.0),
 ('Gage Park', 63.0),
 ('Clearing', 64.0),
 ('West Lawn', 65.0),
 ('Chicago Lawn', 66.0),
 ('West Englewood', 67.0),
 ('Englewood', 68.0),
 ('Greater Grand Crossing', 69.0),
 ('Ashburn', 70.0),
 ('Auburn Gresham', 71.0),
 ('Beverly', 72.0),
 ('Washington Height', 73.0),
 ('Mount Greenwood', 74.0),
 ('Morgan Park', 75.0),
 ("O'Hare", 76.0),
 ('Edgewater', 77.0),
 ('CHICAGO', None)]

### Problem 3

##### List all case numbers for crimes involving minors?(children are not considered minors for the purposes of crime analysis) 


In [10]:
%sql SELECT * FROM CHICAGO_CRIME_DATA WHERE (PRIMARY_TYPE LIKE "%MINORS" OR DESCRIPTION LIKE "%MINOR%")

 * sqlite:///FinalDB.db
Done.


[(3987219, 'HL266884', '2005-03-31', '024XX N CLARK ST', '2210', 'LIQUOR LAW VIOLATION', 'SELL/GIVE/DEL LIQUOR TO MINOR', 'CONVENIENCE STORE', 1, 0, 2333, 19, 43.0, 7.0, '22', 1172680.0, 1916483.0, 2005, 41.92626872, -87.64089934, '(41.926268719, -87.640899336)'),
 (3266814, 'HK238408', '2004-03-13', '093XX S STONY ISLAND AVE', '2230', 'LIQUOR LAW VIOLATION', 'ILLEGAL CONSUMPTION BY MINOR', 'ALLEY', 1, 0, 413, 4, 8.0, 48.0, '22', 1188539.0, 1843379.0, 2004, 41.72530099, -87.58496589, '(41.72530099, -87.584965887)')]

### Problem 4

##### List all kidnapping crimes involving a child?


In [11]:
%sql SELECT * FROM CHICAGO_CRIME_DATA WHERE PRIMARY_TYPE LIKE "%KIDNAPPING%" 

 * sqlite:///FinalDB.db
Done.


[(5276766, 'HN144152', '2007-01-26', '050XX W VAN BUREN ST', '1792', 'KIDNAPPING', 'CHILD ABDUCTION/STRANGER', 'STREET', 0, 0, 1533, 15, 29.0, 25.0, '20', 1143050.0, 1897546.0, 2007, 41.87490841, -87.75024931, '(41.874908413, -87.750249307)')]

### Problem 5

##### List the kind of crimes that were recorded at schools. (No repetitions)


In [12]:
%sql SELECT DISTINCT(PRIMARY_TYPE) FROM CHICAGO_CRIME_DATA WHERE LOCATION_DESCRIPTION LIKE "%SCHOOL%"

 * sqlite:///FinalDB.db
Done.


[('BATTERY',),
 ('CRIMINAL DAMAGE',),
 ('NARCOTICS',),
 ('ASSAULT',),
 ('CRIMINAL TRESPASS',),
 ('PUBLIC PEACE VIOLATION',)]

### Problem 6

##### List the type of schools along with the average safety score for each type.


In [13]:
%sql SELECT DISTINCT("Elementary, Middle, or High School"), AVG(SAFETY_SCORE) FROM CHICAGO_PUBLIC_SCHOOLS GROUP BY "Elementary, Middle, or High School"

 * sqlite:///FinalDB.db
Done.


[('ES', 49.52038369304557), ('HS', 49.62352941176471), ('MS', 48.0)]

### Problem 7

##### List 5 community areas with highest % of households below poverty line


In [14]:
%sql SELECT COMMUNITY_AREA_NAME, PERCENT_HOUSEHOLDS_BELOW_POVERTY FROM CENSUS_DATA ORDER BY PERCENT_HOUSEHOLDS_BELOW_POVERTY DESC LIMIT 5

 * sqlite:///FinalDB.db
Done.


[('Riverdale', 56.5),
 ('Fuller Park', 51.2),
 ('Englewood', 46.6),
 ('North Lawndale', 43.1),
 ('East Garfield Park', 42.4)]

### Problem 8

##### Which community area is most crime prone? Display the coumminty area number only.


In [15]:
%sql SELECT COMMUNITY_AREA_NUMBER, COUNT(*) FROM CHICAGO_CRIME_DATA GROUP BY COMMUNITY_AREA_NUMBER ORDER BY COUNT(*) DESC LIMIT 1

 * sqlite:///FinalDB.db
Done.


[(25.0, 43)]


Double-click **here** for a hint

<!--
Query for the 'community area number' that has most number of incidents
-->


### Problem 9

##### Use a sub-query to find the name of the community area with highest hardship index


In [16]:
%sql SELECT COMMUNITY_AREA_NAME FROM CENSUS_DATA WHERE HARDSHIP_INDEX IN (SELECT MAX(HARDSHIP_INDEX) FROM CENSUS_DATA)

 * sqlite:///FinalDB.db
Done.


[('Riverdale',)]

### Problem 10

##### Use a sub-query to determine the Community Area Name with most number of crimes?


In [20]:
%sql SELECT COMMUNITY_AREA_NUMBER FROM CHICAGO_CRIME_DATA GROUP BY COMMUNITY_AREA_NUMBER ORDER BY COUNT(*) DESC

 * sqlite:///FinalDB.db
Done.


[(25.0,),
 (None,),
 (23.0,),
 (68.0,),
 (29.0,),
 (28.0,),
 (8.0,),
 (71.0,),
 (24.0,),
 (67.0,),
 (66.0,),
 (69.0,),
 (49.0,),
 (6.0,),
 (61.0,),
 (58.0,),
 (26.0,),
 (46.0,),
 (43.0,),
 (31.0,),
 (22.0,),
 (70.0,),
 (44.0,),
 (42.0,),
 (38.0,),
 (30.0,),
 (27.0,),
 (19.0,),
 (16.0,),
 (63.0,),
 (32.0,),
 (2.0,),
 (53.0,),
 (21.0,),
 (1.0,),
 (48.0,),
 (35.0,),
 (15.0,),
 (14.0,),
 (7.0,),
 (76.0,),
 (72.0,),
 (62.0,),
 (51.0,),
 (45.0,),
 (39.0,),
 (5.0,),
 (3.0,),
 (73.0,),
 (65.0,),
 (50.0,),
 (20.0,),
 (17.0,),
 (10.0,),
 (4.0,),
 (77.0,),
 (64.0,),
 (59.0,),
 (56.0,),
 (54.0,),
 (52.0,),
 (41.0,),
 (37.0,),
 (18.0,),
 (11.0,),
 (75.0,),
 (60.0,),
 (55.0,),
 (47.0,),
 (33.0,),
 (12.0,)]

## Author(s)

<h4> Hima Vasudevan </h4>
<h4> Rav Ahuja </h4>
<h4> Ramesh Sannreddy </h4>

## Contribtuor(s)

<h4> Malika Singla </h4>
<h4>Abhishek Gagneja</h4>
<!--
## Change log

| Date       | Version | Changed by        | Change Description                             |
| ---------- | ------- | ----------------- | ---------------------------------------------- |
|2023-10-18  | 2.6     | Abhishek Gagneja  | Modified instruction set |
| 2022-03-04 | 2.5     | Lakshmi Holla     | Changed markdown.                   |
| 2021-05-19 | 2.4     | Lakshmi Holla     | Updated the question                           |
| 2021-04-30 | 2.3     | Malika Singla     | Updated the libraries                          |
| 2021-01-15 | 2.2     | Rav Ahuja         | Removed problem 11 and fixed changelog         |
| 2020-11-25 | 2.1     | Ramesh Sannareddy | Updated the problem statements, and datasets   |
| 2020-09-05 | 2.0     | Malika Singla     | Moved lab to course repo in GitLab             |
| 2018-07-18 | 1.0     | Rav Ahuja         | Several updates including loading instructions |
| 2018-05-04 | 0.1     | Hima Vasudevan    | Created initial version                        |
-->
## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
